src/typechecker.erl

```diff
@@ -2890,6 +2906,106 @@ check_clause(_Env, _ArgsTy, _ResTy, Term) ->
+    Tys.
+
+%% Normalize, refine, revert normalize if no refinement.
+%% May throw no_refinement.
+refine(OrigTy, Ty, TEnv) ->
+    NormTy = normalize(OrigTy, TEnv),
+    case refine_ty(NormTy, normalize(Ty, TEnv), TEnv) of
+        NormTy -> OrigTy;
+        RefTy  -> RefTy
+    end.
+
+%% May throw no_refinement.
+refine_ty(?type(union, Tys), Ty, TEnv) ->
+    normalize(type(union, [refine(U, Ty, TEnv) || U <- Tys]), TEnv);
+refine_ty(?type(tuple, Tys1), ?type(tuple, Tys2), TEnv)
+  when length(Tys1) > 0 ->
+    %% Non-empty tuple
+    RefTys = [refine(Ty1, Ty2, TEnv) || {Ty1, Ty2} <- lists:zip(Tys1, Tys2)],
+    %% {a|b, a|b} \ {a,a} => {a|b, b} | {b, a|b}
+    PossibleTyss = [case RefTy of
+                        Ty          -> [Ty];
+                        ?type(none) -> [Ty];
+                        _           -> [Ty, RefTy]
+                    end || {Ty, RefTy} <- lists:zip(Tys1, RefTys)],
+    [_|TuplesElems] = pick_from_each(PossibleTyss),
+    Tuples = [type(tuple, TupleElems) || TupleElems <- TuplesElems],
+    normalize(type(union, Tuples), TEnv);
+refine_ty(?type(any), _Ty, _) ->
+    throw(no_refinement);
+refine_ty(Ty, ?type(none), _) ->
+    Ty;
+refine_ty({ann_type, _, [_, Ty1]}, Ty2, TEnv) ->
+    refine_ty(Ty1, Ty2, TEnv);
+refine_ty(Ty1, {ann_type, _, [_, Ty2]}, TEnv) ->
+    refine_ty(Ty1, Ty2, TEnv);
+refine_ty(?type(T, A), ?type(T, A), _) ->
+    type(none);
+refine_ty({atom, _, A}, {atom, _, A}, _) ->
+    type(none);
+refine_ty(?type(list, A), ?type(nil), _TEnv) ->
+    type(nonempty_list, A);
+refine_ty({Tag1, _, M} = Ty, {Tag2, _, N}, _TEnv)
+    when Tag1 == integer orelse Tag1 == char,
+         Tag2 == integer orelse Tag2 == char ->
+    if M == N -> type(none);
+       M /= N -> Ty
+    end;
+refine_ty(Ty, {Tag, _, V}, TEnv) when ?is_int_type(Ty),
+                                      Tag == integer orelse Tag == char ->
+    case Ty of
+        ?type(non_neg_integer) when V == 0 -> type(pos_integer);
+        ?type(non_neg_integer) when V < 0  -> Ty;
+        ?type(pos_integer)     when V =< 0 -> Ty;
+        ?type(neg_integer)     when V < 0  -> Ty;
+        ?type(integer)         when V == 0 ->
+            type(union, [type(neg_integer), type(pos_integer)]);
+        ?type(range, [{_TagLo, _, Lo}, {_TagHi, _, Hi}]) ->
+            if
+                Lo =< V, V =< Hi ->
+                    %% It's in the range
+                    Tys = int_range_to_types({Lo, V - 1}) ++
+                          int_range_to_types({V + 1, Hi}),
+                    normalize(type(union, Tys), TEnv);
+                true ->
+                    Ty
+            end;
+        _ ->
+            %% Not possible to subtract from other integer types
+            throw(no_refinement)
+    end;
+refine_ty(Ty1, Ty2, TEnv) ->
+    case subtype(Ty2, Ty1, TEnv) of
+        false     -> Ty1;                 %% disjoint
+        {true, _} -> throw(no_refinement) %% inexact
+    end.
```

Comment:
Return `Ty1` in both cases? See comment about imprecision in integer types and `any()` above.

Reply:
Using `subtype` here is probably fine, but for subtle reasons: `Ty2` will always be very precise, so that if `Ty2` is not a subtype of `Ty1` then they are disjoint. I think it would be more clear, and more future proof, to check if `glb(Ty1, Ty2)` returns `none()`.

Reply:
Sure. I assumed `glb` isn't complete yet. I'll use it if it works; otherwise add a TODO comment.

Reply:
I believe it should work for the types you generate here, but please test 😄.

Reply:
Using glb works.